lib/ecto/repo/preloader.ex

```diff
@@ -154,6 +155,28 @@ defmodule Ecto.Repo.Preloader do
 
   defp preload_assocs([], [], _repo_name, _tuplet), do: []
 
+  defp preload_embeds(structs, [], _repo_name, _tuplet), do: structs
+
+  defp preload_embeds(structs, embeds, repo_name, tuplet) do
+      Enum.map(structs, fn struct ->
+          Enum.reduce(embeds, struct, fn {embed, sub_preloads}, struct_acc ->
```

Comment:
I think you want to collect all the structs together instead of preloading each struct one by one. That's what it looks like is happening for regular assocs.

Reply:
I applied `Task.async_stream` like `preload_assocs` do, is this what you meant?

Reply:
btw is there a way to test for this type of detail? It seems difficult to build a test that makes sure something is happening concurrently.

Reply:
Oh sorry what I meant is that `preload_assocs` calls `preload_each(structs,...)` with the list of relevant structs. But `preload_embeds` is calling `preload_each` one time for each relevant struct.

Reply:
Not sure if I'm making sense but I'll try to explain the downsides

Both ways you've presented so far (Enum.map and Task.async_stream) will perform a different query for each individual struct instead of batching them all into one query. So there is an inefficiency there in terms of # of queries.

There are some other downsides too. The first way (Enum.map) will perform the queries synchronously, hurting performance. The second way (Task.async_stream) will tie up more DB connections than expected. Ecto advertises that it's one extra connection per assoc but now you have multiple connections for a single assoc.

Reply:
@greg-rychlewski oh that is a great explanation, thank you, let me what I can do 👍 
🐈 
⌨️

Reply:
nice catch btw! 🕵️